File: predict.c
/*
 * Neural Network library
 * Copyright (c) 2019-2025 SynthInt Technologies, LLC
 * https://synthint.ai
 * SPDX-License-Identifier: Apache-2.0
 */

#include <stdint.h>
#include <stdio.h>
#include "nn.h"

int main(void) {
  nn_t *model;

  // Test data upon which to make a prediction
  // This is a 16x16 matrix of single bits
  float inputs[] = {
      0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0,
      0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0,
      0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0,
      0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0,
      0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0,
      0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0,
      0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0,
      0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0,
      0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0,
      0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0,
  };
  // Recall a previously trained neural network model, inclusive of its weights
  model = nn_load_model_ascii("model.txt");
  if (NULL == model) {
    printf("Error: Missing or invalid model file.\n");
    return 1;
  }
  // Make an output prediction based upon new input data
  float *prediction = nn_predict(model, inputs);
  for (int i = 0; i < model->width[model->depth - 1]; i++)
    printf("%d: %.5f\n", i, prediction[i]);
  nn_free(model);
  return 0;
}

File: predict.py
#!/usr/bin/env python

# Neural Network library
# Copyright (c) 2019-2025 SynthInt Technologies, LLC
# https://synthint.ai
# SPDX-License-Identifier: Apache-2.0

import sys
from nn import Nn

b = Nn()
b.load('model.txt')
try:
	b.model.contents
except ValueError:
	print('Cannot load model file')
	sys.exit(1)

line = [0,0,0,1,1,1,1,1,0,0,0,0,0,0,0,0.0,
	0,0,1,1,1,1,1,1,1,1,1,0,0,0,0,0.0,
	0,0,1,1,1,0,0,0,1,1,1,0,0,0,0,0.0,
	0,0,1,1,1,0,0,0,1,1,1,0,0,0,0,0.0,
	0,0,1,1,1,0,0,0,1,1,1,0,0,0,0,0.0,
	0,0,0,1,1,0,0,0,1,1,0,0,0,0,0,0.0,
	0,0,0,1,1,1,1,1,0,0,0,0,0,0,0,0.0,
	0,0,0,0,1,1,1,1,0,0,0,0,0,0,0,0.0,
	0,0,0,1,1,1,1,1,1,0,0,0,0,0,0,0.0,
	0,0,1,1,1,1,0,0,1,1,1,0,0,0,0,0.0,
	0,0,1,1,0,0,0,0,0,1,1,1,0,0,0,0.0,
	0,0,1,1,0,0,0,0,0,0,1,1,1,0,0,0.0,
	0,0,1,0,0,0,0,0,0,0,1,1,1,0,0,0.0,
	0,0,1,1,1,0,0,0,0,0,1,1,1,0,0,0.0,
	0,0,0,1,1,1,1,1,1,1,1,1,1,0,0,0.0,
	0,0,0,0,0,0,1,1,1,1,1,1,0,0,0,0.0] 

p = b.predict(line[:b.model.contents.width[0]])
func = lambda x: round(x, 5)
p_list = list(map(func, p[0:10]))
for i in range(len(p_list)):
	print(f'{i}: {p_list[i]:0.5f}')

